### PR TITLE
Arma 3 Egg Rewrite

### PR DESCRIPTION
**Pull request by lilkingjr1**
# Arma 3 Egg Rewrite/Update
### Summary
This is almost a "from the ground up" rewrite and update of my existing Arma 3 egg. It adds many new features and QoL improvements, which are listed below. In general, these include updates to make hosting a server more "user-friendly" and less error prone. In addition, if errors do arise, they are caught more gracefully now, with less cryptic, or more verbose, error messages to the user.

**Please note: This egg will not function properly unless a valid Arma 3 updated image is compiled, hosted, and included in the egg's images list. This PR should not be approved until this occurs!**
___
### Change List:
*In no particular order...*
- Updated all directory READMEs to remove references to old/defunct Arma 3 eggs, fix some title formatting, and fix some links.
- Completely rewrote the Egg's README to...
  - Be in order of information importance
  - Include new Egg Capabilities
  - Include Egg updating instructions
  - Include Recommended Egg Modifications
  - Include grammar and spelling fixes
- Changed the image to the new yolks image.
- Removed the ability to specify custom server.cfg and basic.cfg file locations, because it only complicated things and was rarely changed from their default root location.
- Added the ability to specify a custom basic.cfg URL location for installation (in case it updates in the future).
- Added file parsing of the server.cfg to set the server password, max players, and ensure the headless client IPs are correct.
- Fixed the stop code to actually kill the server
- Adjusted the installation script to pull from the basic.cfg specified URL, and made the SteamCMD login command more robust against credentials with spaces or special characters.
- Rearranged variables to be better grouped/ordered (will only apply if the egg is reinstalled on the panel).
- Rewording of some variables' descriptions.
- Added "Number of SteamCMD Retry Attempts" variable for the image's SteamCMD update function (with new error handling).
- Added "Max Players" variable, so hosts can restrict client's servers if they'd like.
- Added "Server Password" variable to centralize the password used for HCs (if used).
- Added "Disable Mod Downloads/Updates" variable for hosts that do not own the Arma 3 base game and don't want to provide mod downloading support.
- Added "Download Creator DLCs" variable as an easy switch to download CDLCs (people were confused in the Discord on how to run the CDLC "Prairie Fire"; this should help with that).
- Added "Modlist File" variable for defining an Arma 3 Launcher exported HTML modlist file for loading.
- Renamed "Mods" to "Additional Mods".
- Added "Validate Server Files" variable.
- Added "HC Hide Console Output" variable to prevent console spam from HCs.
- Added "Clear HC Profiles Cache on Startup" variable to prevent occasional HC loading issues that can develop over time.
___
### All Submissions:
- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
- [x] Did you branch your changes and PR from that branch and not from your master branch?
___
### Changes to an existing Egg:
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you tested your Egg changes?

### CODE REVIEW

File: game_eggs/steamcmd_servers/arma/README.md
Comment (Software-Noob) [inline game_eggs/steamcmd_servers/arma/README.md:6]: It can be included in the readme when such an egg exists. At this moment, that's unnecessary.

File: game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json
Comment (Software-Noob) [inline game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json:5]: update_url feature does not exist at this moment. It's a placeholder for the future when and if such feature will be implemented

File: game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json
Comment (Software-Noob) [inline game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json:12]: Your image is hosted at `ghcr.io/parkervcp/games:arma3` (Parker has to make the game image packages public first)

File: game_eggs/steamcmd_servers/arma/README.md
Comment (lilkingjr1) [reply game_eggs/steamcmd_servers/arma/README.md]: Agreed. Fixed 33e0e46

File: game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json
Comment (lilkingjr1) [reply game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json:5]: Understood, but does it hurt to be there? I figured it would be good to include just incase the feature is ever implemented. Please reply if you would formally like it removed.

File: game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json
Comment (lilkingjr1) [reply game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json]: That's what I thought it would be too, but @parkervcp [stated in Discord](https://discord.com/channels/122900397965705216/404105683336953866/864613123490316318) it would be `ghcr.io\/parkervcp\/yolks:<tag>`, and I didn't want to go against that. Regardless, fixed: bd311a6

File: game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json
Comment (Software-Noob) [reply game_eggs/steamcmd_servers/arma/arma3/egg-arma3.json:5]: We don't know how it will work, as such this could break it or show a bad example for others looking at the egg

**Comment by pascal-zarrad** [comment]
I really like the new README for the Arma 3 egg. 
However, at the start of your commit history, there are six commits that just have the title `Update README.md` and without opening the description of every single one of them they sound all the same.
I really would love to see them squashed together, I think that would definitly create a cleaner commit history in this case.

**Comment by lilkingjr1** [comment]
@pascal-zarrad Yeh, agreed. Those are all the different directory READMEs within this repo where Arma 3 is mentioned; not just the one main Arma 3 README. I thought people would be looking more at the diffs than at the commits, because all those directory READMEs seemed so inconsequential to give fully detailed titles. However, with them all having the same file name, I can see how it would be confusing to differentiate between each of them. If I knew how to squash them together I would. Maybe you can teach me?

**Comment by lilkingjr1** [comment]
@Software-Noob I'm I still correct to assume that we need to wait for `parkervcp/yolks` to be merged with `pterodactly/yolks` and adjust this egg's image to the new one before we merge this egg? Or are we planning to roll with using Parker's repo/image.

**Comment by Software-Noob** [comment]
> @Software-Noob I'm I still correct to assume that we need to wait for `parkervcp/yolks` to be merged with `pterodactly/yolks` and adjust this egg's image to the new one before we merge this egg? Or are we planning to roll with using Parker's repo/image.

It would use the image from the parkers yolks repo, similar to how the previous Arma 3 image is in the parkers image repo fork of `pterodactyl/images`. The only difference is migration from /images to /yolks. 

The image is built and ready, it only requires review/test to make sure both the egg and image work together as they should before it can be merged

**Comment by pascal-zarrad** [comment]
> @pascal-zarrad Yeh, agreed. Those are all the different directory READMEs within this repo where Arma 3 is mentioned; not just the one main Arma 3 README. I thought people would be looking more at the diffs than at the commits, because all those directory READMEs seemed so inconsequential to give fully detailed titles. However, with them all having the same file name, I can see how it would be confusing to differentiate between each of them. If I knew how to squash them together I would. Maybe you can teach me?

Git has a documentation on how to rewrite the history (changing commit messages, reordering, squashing) here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
You should try out this stuff in a test repository before using it here, as those actions can cause a headache when used without knowing how they work. 

As the different READMEs are for different files, I think you could also just adjust the commit titles to include some indicator telling which README was changed like "Update games README" instead of squashing all of them together.

**Review by Software-Noob — APPROVED** [review]
LGTM, thank you for the extensive contribution!

You can change the readme commits if you want. Personally, they're fine to me in this context, and the commit description explains it in detail - although it's preferable not to use the default Github "Update X file" commit titles for easier quick lookup and git blame.

**Comment by lilkingjr1** [comment]
I'll definitely keep it in mind for next time, but I think I'll have to leave the README commits as it. I can't seem to find a way to do it via the website, I haven't delved into using Git at the command line level yet (I don't think I even have an environment I can use at the moment for that), and I really don't want to mess up the history with some of these more "easily-mistakable" commands.

**Pull request merged**
